### PR TITLE
Tabular: Added efficient OOF functionality to RF/XT models

### PR DESCRIPTION
**Pull request by Innixma**
*Issue #, if available:*

*Description of changes:*

- Added efficient OOF functionality to RF/XT models
- This should give a 10x train/infer speed-up and 10x disk space reduction in bagged modes, with roughly equivalent model quality.
- DAAL and RAPIDS don't support the necessary methods to do this, so only raw sklearn is accelerated at present.
- Fixed efficient OOF KNN crashing on regression problems.

TODO:

- [x] Benchmark

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by szha** [comment]
Job PR-1066-1 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1066/1/index.html

**Comment by Innixma** [comment]
Benchmark Results:

```
                                           framework  > AutoGluon_bestquality_1h_2021_04_15_rf_oof_opt_v2  < AutoGluon_bestquality_1h_2021_04_15_rf_oof_opt_v2  = AutoGluon_bestquality_1h_2021_04_15_rf_oof_opt_v2  % Less Avg. Errors Than AutoGluon_bestquality_1h_2021_04_15_rf_oof_opt_v2  time_train_s  metric_error  time_infer_s  loss_rescaled  time_train_s_rescaled  time_infer_s_rescaled      rank  rank=1_count  rank=2_count  rank=3_count  rank>3_count  error_count
0    AutoGluon_bestquality_1h_2021_04_14_upgrade_cat                                                499                                                  465                                                   55                                            -0.046997                           3403.159470  69921.773190     73.566928       0.463199               1.046596               2.757984  1.483317           500           520             0             0            2
1  AutoGluon_bestquality_1h_2021_04_15_rf_oof_opt_v2                                                  0                                                    0                                                 1021                                             0.000000                           3346.319431  69072.775867     34.807949       0.499509               1.019751               1.185171  1.516683           467           554             0             0            1
```

While we had a slight model quality drop, we reduced average latency by 2.3x and disk usage by likely ~5-10x.

The reasons for the drop are likely due to 2 causes:

1. Only 300 trees are used instead of 3000+ trees for RF and XT models in bagging. This means if RF/XT are the dominant model on a problem, it's likely the old method produces a stronger result.
- While each tree does get more data with the new method, it is unlikely that it will be enough to overcome having 3000+ trees from the old method. This is evident when looking deeper into the results: The new method is stronger on larger datasets because those datasets didn't have time to train the 3000+ trees (or it did but then neglected other models), whereas it is weaker on smaller datasets where repeated bagging often caused RF and XT to have 10000+ trees compared to only 300.
- A possible solution is to train more trees, such as 600, by default.
- Another option is to enable repeated-bagging for this model, adding more trees to the model each time (but not actually bagging).

2. XT is now using bootstrapping, where it previously was not. Bootstrapping is required for the efficient OOF calculation, and this means XT may be weaker in general than before, as it uses less data to train its trees.
- A possible solution would be to re-train XT without bootstrapping and use the bootstrapped version for training, and the non-bootstrapped version for inference.
- Alternatively, adjust the bootstrapping fraction to 10% instead of 38% to get similar strength trees as before. Downside: OOF score will be weaker due to fewer OOB samples. This drawback could be mitigated if we increased num trees to 600.

**Comment by Innixma** [comment]
Overall, despite the slight accuracy drop, I recommend we merge this. The follow-ups can be investigated and even as is, the inference speed and disk usage improvements are enough to make this change worthwhile.

File: core/src/autogluon/core/models/abstract/abstract_model.py [hunk 1]
```diff
@@ -492,7 +497,7 @@ def _convert_proba_to_unified_form(self, y_pred_proba):
                 return y_pred_proba
         elif y_pred_proba.shape[1] > 2:  # Multiclass, Softclass
             return y_pred_proba
-        else:  # Regression
+        else:  # Unknown problem type
```
Comment (jwmueller) [inline core/src/autogluon/core/models/abstract/abstract_model.py:500]: maybe should be marked with TODO?

File: tabular/src/autogluon/tabular/models/rf/rf_model.py [hunk 2]
```diff
@@ -191,6 +195,81 @@ def _predict_proba(self, X, **kwargs):
...
+            normalize = self.normalize_pred_probas
+        if normalize:
+            y_oof_pred_proba = normalize_pred_probas(y_oof_pred_proba, self.problem_type)
+        y_oof_pred_proba = y_oof_pred_proba.astype(np.float32)
+        return y_oof_pred_proba
+
+    # FIXME: Unknown if this works with quantile regression
+    def _get_oof_pred_proba(self, X, y, **kwargs):
+        if self._daal:
+            raise AssertionError('DAAL forest backend does not support out-of-bag predictions.')
```
Comment (jwmueller) [inline tabular/src/autogluon/tabular/models/rf/rf_model.py:211]: These errors may be more helpful if they describe an alternative for the user.

File: tabular/src/autogluon/tabular/models/rf/rf_model.py [hunk 3]
```diff
@@ -191,6 +195,81 @@ def _predict_proba(self, X, **kwargs):
...
+                    self.model.n_classes_ = self.model.n_classes_[0]
+
+        if getattr(self.model, "oob_decision_function_", None) is not None:
+            y_oof_pred_proba = self.model.oob_decision_function_
+            self.model.oob_decision_function_ = None  # save memory
+        elif getattr(self.model, "oob_prediction_", None) is not None:
+            y_oof_pred_proba = self.model.oob_prediction_
+            self.model.oob_prediction_ = None  # save memory
+        else:
+            raise AssertionError(f'Model class {type(self.model)} does not support out-of-fold prediction generation.')
```
Comment (jwmueller) [inline tabular/src/autogluon/tabular/models/rf/rf_model.py:244]: These errors may be more helpful if they describe an alternative for the user.

**Review by jwmueller — APPROVED** [review]
LGTM. I'll check with Taesup regarding possible Quantile Regression functionality issues. 

A remaining concern I have is the user will not understand what special behavior is happening with this RF model / KNN model. It may be worth at least adding a section to the Tabular FAQ page describing this, and possibly explain how the user could instead use regular bagging for these models instead of this special leave-some-out functionality (even though this is not typically recommended).

File: tabular/src/autogluon/tabular/models/rf/rf_model.py [hunk 4]
```diff
@@ -191,6 +195,81 @@ def _predict_proba(self, X, **kwargs):
...
+        if self.problem_type == REGRESSION and self.model.n_estimators <= 40:
+            from sklearn.ensemble._forest import _get_n_samples_bootstrap, _generate_unsampled_indices
+            n_samples = len(y)
+
+            n_predictions = np.zeros(n_samples)
+            n_samples_bootstrap = _get_n_samples_bootstrap(n_samples, self.model.max_samples)
+            for estimator in self.model.estimators_:
+                unsampled_indices = _generate_unsampled_indices(estimator.random_state, n_samples, n_samples_bootstrap)
+                n_predictions[unsampled_indices] += 1
+            missing_row_mask = n_predictions == 0
```
Comment (gradientsky) [inline tabular/src/autogluon/tabular/models/rf/rf_model.py:258]: what about legit zeroes? For example you predict a temperature between -100 and 100?

Comment (gradientsky) [reply tabular/src/autogluon/tabular/models/rf/rf_model.py:258]: Another point: do we want exact 0 here? Float point errors can do this:
```
0.0003 - 0.0002 - 0.0001 == 0
```
This results in `False`

File: core/src/autogluon/core/models/abstract/abstract_model.py (hunk 1 above)
Comment (Innixma) [reply core/src/autogluon/core/models/abstract/abstract_model.py:500]: updated

File: tabular/src/autogluon/tabular/models/rf/rf_model.py [hunk 5]
```diff
@@ -191,6 +195,81 @@ def _predict_proba(self, X, **kwargs):
...
+        elif getattr(self.model, "oob_prediction_", None) is not None:
+            y_oof_pred_proba = self.model.oob_prediction_
+            self.model.oob_prediction_ = None  # save memory
+        else:
+            raise AssertionError(f'Model class {type(self.model)} does not support out-of-fold prediction generation.')
+
+        # TODO: Regression does not return NaN for missing rows, instead it sets them to 0. This makes life hard.
+        #  The below code corrects the missing rows to NaN instead of 0.
+        # Don't bother if >40 trees, near impossible to have missing
+        if self.problem_type == REGRESSION and self.model.n_estimators <= 40:
```
[the remark this answers is not in the document]
Comment (Innixma) [reply tabular/src/autogluon/tabular/models/rf/rf_model.py]: updated to 60 trees, now 1 in 11 billion chance

File: tabular/src/autogluon/tabular/models/rf/rf_model.py (hunk 2 above)
Comment (Innixma) [reply tabular/src/autogluon/tabular/models/rf/rf_model.py:211]: will plan to add more info when DAAL gets further along.

File: tabular/src/autogluon/tabular/models/rf/rf_model.py (hunk 3 above)
Comment (Innixma) [reply tabular/src/autogluon/tabular/models/rf/rf_model.py:244]: Generally user should never see this unless they are doing very advanced things.

File: core/src/autogluon/core/models/abstract/abstract_model.py [hunk 6]
```diff
@@ -492,7 +497,7 @@ def _convert_proba_to_unified_form(self, y_pred_proba):
...
             return y_pred_proba[:, 1]
```
[the remark this answers is not in the document]
Comment (Innixma) [reply core/src/autogluon/core/models/abstract/abstract_model.py]: Added

**Comment by Innixma** [comment]
I've disabled the OOB OOF opt for now until we pin down the impact related to stack-overfitting.
